serverless/serverless
# Add Amazon S3 Transfer Acceleration support

## What did you implement:

I have added support for S3 Transfer Acceleration, which requires one API call for enabling it and one S3-client parameter on request.

Closes #4252

As discussed in the issue, we'd like to enable this functionality only when a command line option such as `--aws-s3-accelerate` is specified. I have partially implemented this (see my TODOs), but I'd need some help to properly read options and eventually yaml config too.

## How did you implement it:

I have added a new step called `enableBucketAcceleration` to the `uploadArtifacts` pipeline.

One transfer acceleration is enabled, we also need to enable a `useAccelerateEndpoint` parameter on the S3 Client for each request. Apparently because of an SDK issue (discussed [here](https://github.com/aws/aws-sdk-js/issues/281#issuecomment-314045508)), I also had to force `signatureVersion = 'v2'`.

## How can we verify it:

This should work with any serverless project when you use the  `--aws-s3-accelerate` command line option. 

To verify that it actually works, I recommend deploying a package of at least 5MB to a remote location (e.g. `ap-southeast-2`). The time difference should be noticeable (e.g. from 20s down to 5s).

For example:

```yaml
service: s3-acceleration-test
package:
  exclude:
    - "**/*"
  include:
    - handler.js
    - huge-file.jpg  # make sure this exists (mine was 22MB)
provider:
  name: aws
  runtime: nodejs6.10
  region: ap-southeast-2  # a randomly selected remote region

functions:
  hello:
    handler: handler.hello  # this is the default Node.js hello world function
```

My results (after deploying the first time to create the deployment stack):

```bash
$ time sls deploy
real	1m38.352s
user	0m13.690s
sys	0m2.761s
```

```bash
$ time sls deploy --aws-s3-accelerate
real	0m57.608s
user	0m13.156s
sys	0m2.556s
```

(which is 40s faster!!)

## Todos:


***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO
